File: JAVA/Problem3.java
import java.util.ArrayList;
import java.util.Arrays;
class Solution{

  public void dfs(int initColor, int[][] vis, int[] delrow, int[] delcol, int[][] image, int row, int col, int newColor){
    vis[row][col] = newColor;
    int n = image.length;
    int m = image[0].length;
    for(int i=0; i<4; i++){
      int newr = delrow[i] + row;
      int newc = delcol[i] + col;
      if(newr>=0 && newr<n && newc>=0 && newc<m && image[newr][newc]==initColor && vis[newr][newc] != newColor){
        vis[newr][newc] = newColor;
        dfs(initColor, vis, delrow, delcol, image, newr, newc, newColor);
      }
    }
  }

  public int[][] floodFill(int[][] image, int sr, int sc, int newColor){
    int initColor = image[sr][sc];
    int[][] vis = new int[image.length][image[0].length];
    // for(int i=0; i<image.length; i++){
    //   vis[i] = Arrays.copyOf(image[i], image[i].length);
    // }
    for (int i = 0; i < image.length; i++) {
        for (int j = 0; j < image[i].length; j++) {
            vis[i][j] = image[i][j];
        }
    }
    int[] delrow = {-1, 0, 1, 0};
    int[] delcol = {0, -1, 0, 1};
    dfs(initColor, vis, delrow, delcol, image, sr, sc, newColor);
    return vis;
  }

  public static void main(String[] args){
    int[][] image = {{1,1,1},{1,1,0},{1,0,1}};
    int sr = 1;
    int sc = 1;
    int newColor = 2;
    Solution obj = new Solution();
    int[][] ans = obj.floodFill(image, sr, sc, newColor);
    System.out.println("old image: ");
    for(int[] row: image){
      for(int val: row){
        System.out.print(val + " ");
      }
      System.out.println();
    }

    System.out.println("-----------------------------------");

    System.out.println("new image: ");
    for(int[] row: ans){
      for(int val: row){
        System.out.print(val + " ");
      }
      System.out.println();
    }
  }
}

File: Python/Problem3.py
from typing import List, Tuple
import copy
class Solution:
    def dfs(self, initColor: int, vis: List[List[int]], delrow: List[int], delcol: List[int], image: List[List[int]], row: int, col: int, newColor: int) -> None:
        vis[row][col] = newColor
        n,m = len(image),len(image[0])
        for i in range(0,4):
            newr = delrow[i] + row
            newc = delcol[i] + col
            if(0<=newr<n and 0<=newc<m and image[newr][newc]==initColor and vis[newr][newc]!=newColor):
                vis[newr][newc] = newColor
                self.dfs(initColor, vis, delrow, delcol, image, newr, newc, newColor)


    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        initColor = image[sr][sc];
        vis = copy.deepcopy(image);
        delrow = [-1, 0, 1, 0]
        delcol = [0, -1, 0, 1]
        self.dfs(initColor, vis, delrow, delcol, image, sr, sc, newColor)
        return vis;

image = [
    [1,1,1],
    [1,1,0],
    [1,0,1]
]

sr = 1
sc = 1
newColor = 2

obj = Solution()
ans = obj.floodFill(image, sr, sc, newColor)

print("old image: ")
# prints out the same way
# for row in image:
#         print(' '.join(map(str, row)))
for i in range(len(image)):
    for j in range(len(image[i])):
        print(image[i][j], end=' ')
    print()

print("----------------")

print("new image")
# for row in ans:
#         print(' '.join(map(str, row)))
for i in range(len(ans)):
    for j in range(len(ans[i])):
        print(ans[i][j], end=' ')
    print()
